### PR TITLE
Clean up and update application metadata files.

### PR DESCRIPTION
**Pull request by igorpeshansky**
## Description
Bring the third-party application `metadata.yaml` files up-to-date with the public docs, in preparation for eventually using them as the source of truth.

Also add some more YAML validation and improve test output.

## Related issue
[b/240700717](http://b/240700717)

## How has this been tested?
Except for the new code, which was tested by unit tests, all of the `metadata.yaml` changes have been run through an automatic doc generator and compared with the existing documentation. Some of the changes also update the scripts that run in the third-party application tests — those have passed.

## Checklist:
- Unit tests
  - [ ] Unit tests do not apply.
  - [x] Unit tests have been added/modified and passed for this PR.
- Integration tests
  - [ ] Integration tests do not apply.
  - [x] Integration tests have been added/modified and passed for this PR.
- Documentation
  - [x] This PR introduces no user visible changes.
  - [ ] This PR introduces user visible changes and the corresponding documentation change has been made.
- Minor version bump
  - [x] This PR introduces no new features.
  - [ ] This PR introduces new features, and there is a separate PR to bump the [minor version](https://github.com/GoogleCloudPlatform/ops-agent/blob/master/VERSION) since the last [release](https://github.com/GoogleCloudPlatform/ops-agent/releases) already.
  - [ ] This PR bumps the version.

### CODE REVIEW

File: integration_test/third_party_apps_test/applications/apache/metadata.yaml [hunk 1]
```diff
@@ -25,8 +25,7 @@ description: |-
 configure_integration: |-
   To collect telemetry from your Apache Web Server, you must configure the
   server's `httpd.conf` file to enable the
-  [`mod_status` plugin](https://httpd.apache.org/docs/2.4/mod/mod_status.html)
-  {: .external}.
+  [`mod_status` plugin](https://httpd.apache.org/docs/2.4/mod/mod_status.html){: .external}.
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/apache/metadata.yaml:28]: Delete?

File: integration_test/third_party_apps_test/applications/elasticsearch/metadata.yaml [hunk 2]
```diff
@@ -26,10 +26,9 @@ configure_integration: |-
   If you enable [Elasticsearch security
   features](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/secure-cluster.html){: .external},
   you must configure a user with the `monitor` or `manage` [cluster
-  privilege](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/security-privileges.html#privileges-list-cluster)
-  {: .external}.
+  privilege](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/security-privileges.html#privileges-list-cluster){: .external}.
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/elasticsearch/metadata.yaml:29]: Delete?

File: integration_test/third_party_apps_test/applications/jetty/metadata.yaml [hunk 3]
```diff
@@ -21,11 +21,32 @@ description: |-
   The Jetty integration collects session and thread-usage metrics. The
   integration also collects access and system logs. Access logs are parsed
   into a JSON payload focused on request details.
+configure_integration: |-
+  To collect metrics from this integration, you must expose a
+  [Java Management Extensions (JMX)](https://www.oracle.com/java/technologies/javase/javamanagement.html){:class="external"}
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/jetty/metadata.yaml:26]: Delete?

File: integration_test/third_party_apps_test/applications/mssql/metadata.yaml [hunk 4]
```diff
@@ -17,7 +17,19 @@ app_url: "https://docs.microsoft.com/en-us/sql/sql-server"
...
+  instances. The metrics provide transaction-rate information and also data on
+  open connections.
+
+  The v2 Microsoft SQL Server integration collects metrics from your SQL Server
+  instances and databases. Page, batch, lock, and user connection count information
+  is provided at the instance level. Transaction and transaction log information
+  is provided at the database level.
+
+  For more information on SQL Server, see the
+  [SQL Server](https://docs.microsoft.com/en-us/sql/sql-server){: .external}
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/mssql/metadata.yaml:31]: Delete?

File: integration_test/third_party_apps_test/applications/varnish/metadata.yaml [hunk 5]
```diff
@@ -21,10 +21,18 @@ description: |-
   It monitors the number of objects entering and exiting the cache,
   as well as the number of sessions and backend connections.
   The integration also collects Varnish logs and parses them
-  into a standardized json payload.
+  into a standardized JSON payload.
 supported_app_version: ["6.x", "7.0.x"]
 configure_integration: |-
-  The Varnish logging processor processes logs using [varnishncsa](https://varnish-cache.org/docs/6.0/reference/varnishncsa.html){: .external}. Varnish can enable varnishncsa logging by following [this](https://docs.varnish-software.com/tutorials/enabling-logging-with-varnishncsa/){: .external} guide depending on os. By default, logs write to /var/log/varnish/varnishncsa.log. If another file is chosen, the user will have to also update the receiver configuration. Logs are expected to be in the default format and a logrotation should be setup.
+  The Varnish logging processor processes logs using
+  [varnishncsa](https://varnish-cache.org/docs/6.0/reference/varnishncsa.html){: .external}.
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/varnish/metadata.yaml:28]: Delete?

File: integration_test/third_party_apps_test/applications/varnish/metadata.yaml [hunk 6]
```diff
@@ -21,10 +21,18 @@ description: |-
...
+  Varnish can enable varnishncsa logging by following this [guide](https://docs.varnish-software.com/tutorials/enabling-logging-with-varnishncsa/){: .external},
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/varnish/metadata.yaml:29]: Delete?

File: integration_test/third_party_apps_test/applications/elasticsearch/metadata.yaml [hunk 7]
```diff
@@ -26,10 +26,9 @@ configure_integration: |-
   If you enable [Elasticsearch security
   features](https://www.elastic.co/guide/en/elasticsearch/reference/7.16/secure-cluster.html){: .external},
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/elasticsearch/metadata.yaml:27]: Delete?

File: integration_test/third_party_apps_test/applications/mariadb/metadata.yaml [hunk 8]
```diff
@@ -36,8 +36,11 @@ configure_integration: |-
   shown in the table below. On some platforms, MariaDB logs to `journald` by
   default instead of a file. Configure MariaDB to log to a file instead by
   setting `log_error` in the MariaDB configuration.
-  [Learn more](https://mariadb.com/kb/en/error-log/#writing-the-error-log-to-a-file)
+  [Learn more](https://mariadb.com/kb/en/error-log/#writing-the-error-log-to-a-file){: .external}
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/mariadb/metadata.yaml:39]: Delete?

File: integration_test/third_party_apps_test/applications/tomcat/metadata.yaml [hunk 9]
```diff
@@ -23,7 +23,7 @@ description: |-
   access and Catalina logs. Access logs are parsed into a JSON payload
   focused on request details, whereas Catalina logs are parsed for general
   details. The `tomcat` receiver collects telemetry from the Tomcat server's Java
-  Virtual Machine (JVM) via [JMX](https://www.oracle.com/java/technologies/javase/javamanagement.html).
+  Virtual Machine (JVM) via [JMX](https://www.oracle.com/java/technologies/javase/javamanagement.html){: .external}.
```
Comment (mtabasko) [inline integration_test/third_party_apps_test/applications/tomcat/metadata.yaml:26]: Delete?

**Review by mtabasko — APPROVED** [review]
Just flagged a bunch of "external" markers. Would be nice to clean up, but it's hardly a show-stopper. This looks so much better that it did. Thank you!

File: integration_test/third_party_apps_test/applications/activemq/metadata.yaml [hunk 10]
```diff
@@ -18,14 +18,18 @@ short_name: ActiveMQ
...
-  The Apache ActiveMQ integration collects storage usage and message metrics.
-  Storage metrics include memory and disk usage. Message metrics include number of
-  waiting messages, average wait time, and expired messages.
+  The Apache ActiveMQ integration writes logs and collects storage usage and
+  message metrics. Storage metrics include memory and disk usage.
+  Message metrics include number of waiting messages, average wait time,
+  and expired messages.
+configure_integration: |-
+  You must enable JMX support in the ActiveMQ [broker
+  configuration](https://activemq.apache.org/jmx){:class="external"}.
```
[the remark this answers is not in the document]
Comment (igorpeshansky) [reply integration_test/third_party_apps_test/applications/activemq/metadata.yaml:27]: Definitely plan to delete them in the next iteration. Step 1 was to bring this to a minimal diff with the current docs, and those docs have the external marker here. Alternatively, we can delete them from the current docs, and not add them here. Let me work on the latter, then.

[the remark this answers is not in the document]
Comment (igorpeshansky) [reply integration_test/third_party_apps_test/applications/activemq/metadata.yaml]: On second thought, sent a change to delete them internally, and blew them away in this PR.

File: integration_test/metadata/integration_metadata.go [hunk 11]
```diff
@@ -121,13 +121,16 @@ type IntegrationMetadata struct {
...
-func UnmarshalAndValidate(data []byte, i interface{}) error {
+func UnmarshalAndValidate(fullPath string, data []byte, i interface{}) error {
 	data = bytes.ReplaceAll(data, []byte("\r\n"), []byte("\n"))
 
 	v := NewIntegrationMetadataValidator()
 	// Note: Unmarshaler does not protect when only the key being declared.
 	// https://github.com/goccy/go-yaml/issues/313
-	return yaml.UnmarshalWithOptions(data, i, yaml.DisallowUnknownField(), yaml.Validator(v))
+	if err := yaml.UnmarshalWithOptions(data, i, yaml.DisallowUnknownField(), yaml.DisallowDuplicateKey(), yaml.Validator(v)); err != nil {
+		return fmt.Errorf("%s%w", fullPath, err)
```
Comment (jefferbrecht) [inline integration_test/metadata/integration_metadata.go:131]: Nit: can we unsquish the `%s` and the `%w`? Or will they be implicitly separated by a newline or something?

File: integration_test/validate_metadata_test.go [hunk 12]
```diff
@@ -61,18 +69,19 @@ func TestRequireMetadataForAllThirdPartyApps(t *testing.T) {
...
+			integrationMetadata := &metadata.IntegrationMetadata{}
+			err := metadata.UnmarshalAndValidate(fullPath, contents, integrationMetadata)
+			if integrationMetadata.PublicUrl == "" {
+				// The public doc isn't available yet.
+				return
+			}
+			if err != nil {
+				t.Error(err)
+			}
 			r, err := http.Get(integrationMetadata.PublicUrl)
```
Comment (jefferbrecht) [inline integration_test/validate_metadata_test.go:85]: We had a recent flake here with 5xx errors failing the test. Is this a good opportunity to either add retries or to move it to an integration test?

File: integration_test/metadata/integration_metadata.go (hunk 11 above)
Comment (igorpeshansky) [reply integration_test/metadata/integration_metadata.go:131]: No, this is intentional. The YAML parser [reports the position of the error](https://github.com/goccy/go-yaml/blob/master/internal/errors/error.go#L187) in the data as `[<line>:<column>]` at the start of the error. We are prepending the filename to that, so the position immediately follows the filename with no spaces.

File: integration_test/validate_metadata_test.go (hunk 12 above)
Comment (igorpeshansky) [reply integration_test/validate_metadata_test.go:85]: As far as I can see, adding retries properly is a can of worms (what conditions should we retry? immediately or with backoff? how many times/how long? implement your own on top of `net/http` or add a dependency on a more sophisticated library like https://github.com/hashicorp/go-retryablehttp?).  Let's do that in a separate PR?

**Review by jefferbrecht — APPROVED** [review]

**Pull request merged**